src/slotmap.jl

```diff
@@ -0,0 +1,230 @@
+        slots = Memory{UInt64}(undef, startlen)
+        # Start with generation 0
+        slots .= UInt64(1):UInt64(startlen)
+        setfield!(isv, :slots, slots)
+        setfield!(isv, :slots_len, startlen)
+    end
+    @inbounds old_slot = slots[slot_idx]
+    removei! = a -> remove!(a, i)
+    unrolled_map(removei!, values(comps))
+    # Update free linked list
+    free_head = getfield(isv, :free_head)
+    # Free head is zero if the free list is empty
+    if old_slot < gen_mask(isv)
+        # The slot has not been used too many times to cause a generation overflow.
+        @inbounds slots[slot_idx] = UInt64(1)<<63 | old_slot & ~val_mask(isv) | free_head%UInt64
+        setfield!(isv, :free_head, slot_idx)
+    else
+        # Avoid adding the slot back to the free list if it has been used too many times
+        @inbounds slots[slot_idx] = ~UInt64(0)
+    end
+    if i ≤ length(ID)
+        # adjust values in slots because the components have swapped
+        @inbounds moved_pid = ID[i]
+        moved_slot_idx = (moved_pid & val_mask(isv))%Int
+        @inbounds slots[moved_slot_idx] = slots[moved_slot_idx] & ~val_mask(isv) | i
+    end
+    return isv
+end
+
+function Base.deleteat!(isv::SlotMapStructVector, i::Int)
+    comps = getfield(isv, :components)
+    ID = getfield(comps, :ID)
+    delete_id_index!(isv, ID[i], i)
+end
+
+function Base.delete!(isv::SlotMapStructVector, id::Int)
+    i = id_guess_to_index(isv, id, id)
+    delete_id_index!(isv, id, i)
+end
+
+function Base.delete!(isv::SlotMapStructVector, a::IndexedView)
+    id, lasti = getfield(a, :id), getfield(a, :lasti)
+    i = id_guess_to_index(isv, id, lasti)
+    delete_id_index!(isv, id, i)
+end
+
+function Base.push!(isv::SlotMapStructVector, t::NamedTuple)
+    comps, slots = getfield(isv, :components), getfield(isv, :slots)
+    slots_len, free_head = getfield(isv, :slots_len), getfield(isv, :free_head)
+    fieldnames(typeof(comps))[2:end] !== keys(t) && error("Tuple fields do not match container fields")
+    ID = getfield(comps, :ID)
+    startlen = Int64(length(ID))
+    if startlen ≥ val_mask(isv)
+        error("SlotMapStructVector can store at most $(val_mask(isv)) elements.")
+    end
+    if iszero(slots_len)
+        push!(ID, startlen + 1)
+        setfield!(isv, :last_id, startlen + 1)
+    else
+        if iszero(free_head)
+            # push a slot to the end of slots
+            old_slots_capacity = length(slots)
+            if old_slots_capacity ≤ slots_len
+                if old_slots_capacity ≥ val_mask(isv)
+                    error("SlotMapStructVector is out of capacity")
+                end
+                # reallocate the slots
+                # avoid having a capacity larger than val_mask(isv)
+                new_slots_capacity = clamp(
+                    overallocation(old_slots_capacity),
+                    old_slots_capacity+1,
+                    val_mask(isv)
+                )
```

Comment:
```suggestion
                new_slots_capacity = min(overallocation(old_slots_capacity), val_mask(isv))
```

Reply:
Yes, looking at this again, I think the only way for this to overflow is if this is running with 32-bit Julia